### ISSUE TITLE
Elmish Example

### ISSUE THREAD

**Issue by ccorcos** [paldepind/flyd#68]
Hey, I just wanted to point out that I'm using this library with huge success. I'm implementing the elm architecture pattern in coffeescript and this package does a great job at replacing / mimicking the elm signals package. I tried using RxJS and it was a nightmare -- I never got it to work. So great job with this package! :+1: 

https://github.com/ccorcos/elmish
https://github.com/ccorcos/elmish/blob/master/elmish.coffee

**Comment by ericgj** [comment]
Hey, cool stuff! Coincidentally I just did a [flyd + ramda translation of StartApp](https://gist.github.com/ericgj/d373d7e5d7131545938a) too. Great to see others exploring the same space. 

Re `mergeAll`, I found that `R.commute` does the trick, since flyd streams implement `map` and `ap`.

I hadn't considered modelling effects as a stream, but I guess that makes sense with promises - since flyd flattens/resolves promises automatically. That makes things easy!

PS. in case you haven't seen it, there is a repo Simon set up collecting examples of the Elm architecture in Javascript. You might want to send a pull request to add some of your examples there:  https://github.com/paldepind/functional-frontend-architecture

**Comment by ccorcos** [comment]
haha looks like im not alone! thanks for the links! :+1: 

hmm, idk if commute would really work. Else I'd have a stream of streams...

I'm not entirely sure about modeling effects as a stream. I started with views returning effects as a stream and merging them all the way up. But then the start app was less pure because I had use `flyd.to` to pipe the combined effects to the messages stream. I realized though, that in Elm, concurrency needs to be managed with a port. But in javascript, you dont have to worry about that so its pretty convenient just to pass the effects stream all the way down with forwardTo. Because, in javascript, what would be the equivalent of Effects, Task and port? Task is kind of like a promise that doesnt run immediately. Its just an async function really...

**Comment by ericgj** [comment]
Commute does what your mergeAll does, but in a more generic way. I haven't tested it thoroughly, but that's what I understand from the type signatures, and it seems to work. Yours might be more efficient though.

Re Effects, I see what you mean about passing the effects stream down (in `init` and `update`), although I don't see where you use `forwardTo` ?   Also, what if you needed to perform a side effect from a nested component -- how would that component get the effects stream?  It seems like it would need to originate from a parent update passing down the effects stream to a child action. But sometimes you have situations where the child component originates the action. So maybe you should pass `effect$`, like `action$`, down through `view` and `init` ?

The way I deal with effects (which I got from @paldepind, and is close to the way they flow in Elm, except Elm has a special `Effects` wrapper around Tasks), rather than having components place them directly on a stream, is to return them from update functions as tasks (similar to promises). So update functions return `[newmodel, [task1, task2, ...]]`  (note you can trigger multiple tasks from a single action). Only at the top level (in startapp), is where the tasks are actually executed and their results passed into the action stream. (There are some examples of this, including with nested components, in the repo I linked to above.)

I imagine you could do something similar with promises, i.e. at the top level pull out the promises from your model stream, and throw their results into the action stream. Something like this (simplified):

```coffeescript
modelAndEffect$  = flyd.scan( update, init(), action$)
model$ = flyd.map(R.head, modelAndEffect$)
effect$ = flyd.map(R.nth(1), modelAndEffect$)
flyd.stream([effect$], action$)   # since effect$ resolves the promises
```

**Comment by ccorcos** [comment]
I use forwardTo [here](https://github.com/ccorcos/elmish/blob/master/listOf.coffee#L20) for example and I map it over [toChildAction](https://github.com/ccorcos/elmish/blob/master/listOf.coffee#L4) so the action can originate at the child and make its way back to it. 

The dispatch$ (passed to the view) and the effect$ (passed to init and update) are effectively the same thing. They're different streams only so I could perhaps analyze them separately for some reason. But they are simply merged with the inputs into a single action$ [here](https://github.com/ccorcos/elmish/blob/master/elmish.coffee#L37-L43).

Thats actually the pattern I originally used returning [an array of promises](https://github.com/ccorcos/elmish/blob/96ab7b9769a3c62001c265c5740305983f0e7a81/giphy.coffee#L26) but I found it [really ugly to map over](https://github.com/ccorcos/elmish/blob/96ab7b9769a3c62001c265c5740305983f0e7a81/list.coffee#L12-L14). Then I realized that the promises call the async functions immediately anyways. So if I wanted to execute them only at the top-level, then I'd need to create my own a functor I could map over which I could then execute and the action will be nested appropriately. This just seems like a lot of work for no gain though. So I just thought I'd pass the stream directly... One benefit is you get a really clean implementation without having to [do something like this](https://gist.github.com/ericgj/d373d7e5d7131545938a#file-startapp-js-L77).  

Also one of the things I'm confused about is how [this array](https://gist.github.com/ericgj/d373d7e5d7131545938a#file-startapp-js-L55) doesnt compound to infinity...

**Comment by ericgj** [comment]
> The dispatch$ (passed to the view) and the effect$ (passed to init and update) are effectively the same thing. They're different streams only so I could perhaps analyze them separately for some reason. But they are simply merged with the inputs into a single action$ here.

It occurs to me you could even get away with a single action$ (or dispatch$) stream containing _either_ actions _or_ promises resolving to actions, and flyd would deal with them either way. The beauty (and/or horror) of untyped javascript!!

> So I just thought I'd pass the stream directly... One benefit is you get a really clean implementation without having to do something like this.

What I am doing in that line is putting all my side effects in one place, at the top level, and minimizing places in my components where I imperatively put stuff on streams. Also it's needed since Tasks (unlike promises) do not automatically execute until you `fork` them.

> Also one of the things I'm confused about is how this array of accumulated effects doesnt compound to infinity...

Because the context it's called is a single update, where [the array is initialized empty](https://gist.github.com/ericgj/d373d7e5d7131545938a#file-startapp-js-L59-L61). 

Basically you have an action stream which has "the list of actions at the current moment from all inputs". Each of these actions returns `[state, [effects]]` when fed through the top-level component update function. You want to process each of these actions sequentially, using the top-level component update to _reduce_ the state while _accumulating_ a list of the returned effects (to be pulled out and executed downstream).  That's what this function is doing.

**Comment by ccorcos** [comment]
> The beauty (and/or horror) of untyped javascript!!

![screen shot 2015-09-10 at 10 06 55 am](https://cloud.githubusercontent.com/assets/1794527/10807741/3daf59a0-7da0-11e5-8ce6-bb02c887b6cf.png)

hahahaha

> What I am doing in that line is putting all my side effects in one place, at the top level, and minimizing places in my components where I imperatively put stuff on streams. 

I do like this. And I like how you're using Futures. What would be better is if these were serializable actions though and they were processed elsewhere, like an http service. This way you can inspect them and analyze them. You could batch tasks together when appropriate and send only one http request as opposed to several... just a thought.

> You want to process each of these actions sequentially, using the top-level component update to reduce the state while accumulating a list of the returned effects (to be pulled out and executed downstream). That's what this function is doing.

I see.

At this point, my goal is to figure out publish/subscribe over websockets and make a simple chat client. Hopefully it all works out. Ideally the actions could be sent directly into the websocket -- how cool would that be? Perhaps the server can be modelled in the exact same way... we'll see. I plan on asking the elm wizards tn ;)

thanks again for all the help -- much appreciated.

**Comment by ericgj** [comment]
> I do like this. And I like how you're using Futures. What would be better is if these were serializable actions though and they were processed elsewhere, like an http service. This way you can inspect them and analyze them. You could batch tasks together when appropriate and send only one http request as opposed to several... just a thought.

It's interesting - in principle this is possible. At the component level your tasks could be defined in terms of some kind of "remote action type", then at the top level merge these into a single task chain that optimizes the http requests etc. and also sorts out how to distribute the results back down to actions. I'm sure someone is doing this kind of thing. 

I would drop a line on the [ramda gitter room](https://gitter.im/ramda/ramda), or actually [fantasy-land gitter](https://gitter.im/fantasyland/fantasy-land) has had a bit of chatter about Futures recently. Probably a better place than this issue thread on flyd ;)
